enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,439 @@
+that implements additional pool.
+When shared-partition lands, Kubelet's cpu pool layout should look roughly like:
+1. `reservedSystemCpus`
+2. `guaranteedCpus` <- new pool
+3. `sharedCpus` = All CPUs - `reservedSystemCpus` - `guaranteedCpus`
+
+When shared-partition feature enabled, 
+mixed-cpu-node-plugin allocates the shared cpus from the `sharedCpus` pool and not
+from the `reservedSystemCpus` anymore.
+
+The following table shows the shared-cpus origin in each flow:
+
+| mixed-cpu | shared-partition | shared cpus origin                 |
+|-----------|------------------|------------------------------------|
+| Enable    | Disable          | Kubelet's `reservedSystemCPUs`     | 
+| Disable   | Enable           | All cpus - `reserved` - `isolated` | 
+| Enable    | Enable           | All cpus - `reserved` - `isolated` | 
+
+#### Workload configuration
+A workload that wants to access the shared cpu should request for `openshift.io/enabled-shared-cpus` under its spec.
+The `openshift.io/enabled-shared-cpus` should be treated as a boolean value.
+In other words, the resource request only uses as a hint that shared-cpus required for the container and
+does not indicate any actual value.
+
+For example:
+```yaml
+requests:
+   openshift.io/enabled-shared-cpus: 1
+```
+Fine.
+
+```yaml
+requests:
+   openshift.io/enabled-shared-cpus: 2
+```
+Wrong. An error will be returned to the user explaining how to fix the pod spec.
+
+The reason for specifying a resource request, and not an annotation,
+is to allow control on the number of workloads that can request for shared cpus.
+By populating shared cpu resources as the number of maximum workloads.
+If the number of shared cpu requests got exhausted, any new workload that requests for shared cpus request will move
+to pending state by the scheduler.
+
+Once workload request completed successfully it gets access to shared cpus, and two new environment variables
+will be present under the container's environment variables:
+`OPENSHIFT_SHARED_CPUS=<CPU-IDs>` = specifies the core ids of the shared cpus
+`OPENSHIFT_ISOLATED_CPUS=<CPU-IDs>` = specifies the core ids of the isolated cpus
+Those environment variables help the application's user to pin its processes/threads to the desired CPU set. 
+
+#### Kubelet
+Kubelet needs to be updated to advertise the `openshift.io/enabled-shared-cpus` resources 
+as [Extended Resources](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#extended-resources).
+Kubelet will look for a configuration file `/etc/kubernetes/openshift-shared-cpus`, to enable the resource advertisement.
+The configuration file should be look like this (the number of resources might be varied)
+``` json
+{
+  "shared-cpus": {
+     "quantity": "15"
+  }
+}
+```
+
+#### API Server Admission Hook
+A new admission hook for [OpenShift Kubernetes API Server](https://github.com/openshift/kubernetes/tree/master/openshift-kube-apiserver/admission)
```

Comment:
Can we fold this logic  into the mgmt core pinning admission hook?

Reply:
@mrunalp I think so, do you have a link for the existing logic?

Reply:
Here is the code for that - https://github.com/openshift/kubernetes/commit/04ff5090bae1cb181a2464696adde8709cdd0a93